x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleRunner.java

```diff
@@ -340,34 +393,65 @@ static ClusterState moveClusterStateToErrorStep(Index index, ClusterState cluste
         causeXContentBuilder.startObject();
         ElasticsearchException.generateThrowableXContent(causeXContentBuilder, STACKTRACE_PARAMS, cause);
         causeXContentBuilder.endObject();
-        LifecycleExecutionState nextStepState = moveExecutionStateToNextStep(policyMetadata,
-            LifecycleExecutionState.fromIndexMetadata(idxMeta), currentStep, new StepKey(currentStep.getPhase(),
-                currentStep.getAction(), ErrorStep.NAME), nowSupplier, false);
+        LifecycleExecutionState currentState = LifecycleExecutionState.fromIndexMetadata(idxMeta);
+        LifecycleExecutionState nextStepState = moveExecutionStateToNextStep(policyMetadata, currentState, currentStep,
+            new StepKey(currentStep.getPhase(), currentStep.getAction(), ErrorStep.NAME), nowSupplier, false);
         LifecycleExecutionState.Builder failedState = LifecycleExecutionState.builder(nextStepState);
         failedState.setFailedStep(currentStep.getName());
         failedState.setStepInfo(BytesReference.bytes(causeXContentBuilder).utf8ToString());
+        Step failedStep = stepLookupFunction.apply(idxMeta, currentStep);
+        // as an initial step we'll mark the failed step with a transitive error without actually looking at the cause to determine
+        // if it's a transitive (ie. recoverable from) error
+        failedState.setIsTransitiveError(failedStep.isRetryable());
+        // maintain the retry count of the failed step as it will be cleared after a successful execution
+        failedState.setFailedStepRetryCount(currentState.getFailedStepRetryCount());
+
         ClusterState.Builder newClusterStateBuilder = newClusterStateWithLifecycleState(index, clusterState, failedState.build());
         return newClusterStateBuilder.build();
     }
 
+    ClusterState moveClusterStateToRetryFailedStep(ClusterState currentState, String index) {
+        return moveClusterStateToFailedStep(currentState, index, true);
+    }
+
+    private ClusterState moveClusterStateToFailedStep(ClusterState currentState, String index, boolean isRetry) {
```

Comment:
I think it's getting confusing having `moveClusterStateToErrorStep` and `moveClusterStateToFailedStep`. The names are very similar and it's difficult to tell at a glance how the two differ.

How about we rename this to something like `moveClusterStateOutOfErrorStep` so it makes it clearer that one is entering the error state and the other is exiting the error state?

Reply:
Also, the `isRetry` is an overloaded term since we have a "retry" API, maybe it should be `isAutomaticRetry`?

Reply:
Thinking about this rename some more, it seems to me that `moveClusterStateToFailedStep` is actually more descriptive than `moveClusterStateOutOfErrorStep` (seeing a method like this, I'd immediately ask myself "moving where?" and I'd have to look at what it does to answer that question)

I appreciate the number of `moveClusterStateToX` methods increased with this PR. The methods we currently have are `moveClusterStateTo`: 
* `FailedStep`
* `ErrorStep`
* `NextStep`
* `Step`
* `RetryFailedStep` - added in this PR but it's just an overload which I'll drop as there's no need for an extra term in the vocabulary

Unfortunately, I don't have a better suggestion to clean up the namings (probably because I've spent a lot of time recently in this code and it makes sense to me) so if `moveClusterStateOutOfErrorStep` makes more sense to you I'll rename it, just wanted to express my concerns before doing so. @dakrone @gwbrown